modules/product/src/VariationFieldInjection.php

```diff
@@ -0,0 +1,171 @@
+   *
+   * @var \Drupal\Core\Field\FieldDefinitionInterface[]
+   */
+  protected $fieldDefinitions = [];
+
+
+  /**
+   * Constructs a new VariationFieldInjector object.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
+   *   The entity type manager.
+   * @param \Drupal\Core\Entity\EntityFieldManagerInterface $entity_field_manager
+   *   The entity field manager.
+   * @param \Drupal\Core\Extension\ModuleHandlerInterface $module_handler
+   *   The module handler.
+   */
+  public function __construct(EntityTypeManagerInterface $entity_type_manager, EntityFieldManagerInterface $entity_field_manager, ModuleHandlerInterface $module_handler) {
+    $this->entityTypeManager = $entity_type_manager;
+    $this->entityFieldManager = $entity_field_manager;
+    $this->moduleHandler = $module_handler;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getFieldDefinitions($variation_type_id) {
+    if (!isset($this->fieldDefinitions[$variation_type_id])) {
+      $definitions = $this->entityFieldManager->getFieldDefinitions('commerce_product_variation', $variation_type_id);
+      $definitions = array_filter($definitions, function ($definition) {
+        /** @var \Drupal\Core\Field\FieldDefinitionInterface $definition */
+        $name = $definition->getName();
+        if ($definition instanceof BaseFieldDefinition && !in_array($name, $this->allowedBaseFields())) {
+          return FALSE;
+        }
+        return (strpos($name, 'attribute_') === FALSE);
+      });
+
+      $this->fieldDefinitions[$variation_type_id] = $definitions;
+    }
+
+    return $this->fieldDefinitions[$variation_type_id];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getRenderableFields(ProductVariationInterface $variation, $view_mode = 'default') {
+    $injectable_fields = [];
+
+    $field_definitions = $this->getFieldDefinitions($variation->bundle());
+
+    $view_display = $this->entityTypeManager
+      ->getStorage('entity_view_display')
+      ->load("commerce_product_variation.{$variation->bundle()}.$view_mode");
+
+    if (!$view_display) {
+      // Use default if one was not available.
+      /** @var \Drupal\Core\Entity\Entity\EntityViewDisplay $view_display */
+      $view_display = commerce_get_entity_display('commerce_product_variation', $variation->bundle(), 'view');
+    }
+
+    $view_builder = $this->entityTypeManager->getViewBuilder('commerce_product_variation');
+
+    foreach ($field_definitions as $field_name => $field_definition) {
+      if ($view_display->getComponent($field_name)) {
+        $content = $view_builder->viewField($variation->$field_name, $view_mode);
+        $content['#attributes']['class'][] = $this->baseCssClass();
+
+        $content['#attributes']['class'][] = $this->variationFieldCssClass('variation_' . $field_name, $variation);
+        $content['#ajax_injection_class'] = $this->variationFieldCssClass('variation_' . $field_name, $variation);
```

Comment:
very nitpicky and possibly a matter of taste, as the function call is a cheap one, but I'd prefer call the function only once, and re-use the variable instead